### ISSUE TITLE
Makeflow plus Scientific Filesystem proof-of-concept

### ISSUE THREAD

**Issue by julianpistorius** [AgPipeline/computing-pipeline#132]
Proof-of-concept of:

- Canopy cover workflow
    - 'The sequence of the overall workflow is: odm_workflow -> soil_mask_workflow -> plot_clip_workflow -> canopy_cover_workflow'
- Including Open Drone Map
- Single Scientific Filesystem container image (multiple entry points - one for each transformer)
- Single .jx workflow at the top level
- Potentially .jx sub-workflows (if this now works in Makeflow) or shell scripts which run other .jx workflows
    - Chris says should work, as long as sub-workflows are not in a loop
- Workflow paramaterized via standard Makeflow variables file
- Using zip file from Chris with a small subset of test data (drone photography over gantry field)
    - <https://drive.google.com/file/d/1paeHs2Kd2ib2Maces9sTC-jFxV4tm8io/view?usp=sharing>
- Reproducing functionality currently in: <https://github.com/Chris-Schnaufer/drone-makeflow>

## Tasks

- [x] Package two transformers in one container using Scientific Filesystem #133
- [x] Makeflow to run two transformers using one container #134
  - Narrowed scope to just getting Makeflow working, and created follow up as #137 
 - [x] Call two SCIF transformers using Makeflow #137
- [x] All transformers for canopy cover workflow in one container
- [ ] Makeflow for entire canopy cover workflow using one container
- [ ] Cookiecutter recipe for new transformer
- [ ] Cookiecutter recipe for new workflow

## Stretch goals

- [ ] Submit a Makeflow recipe to https://github.com/sci-f/apps
  - See https://sci-f.github.io/community
- [ ] Submit a SCIF example https://github.com/cooperative-computing-lab/makeflow-examples

## Questions

- [ ] Unique folder per Makeflow run - Can Makeflow do this automatically?
    - Dynamic environment variables?
    - Julian to investigate
- [x] Do we have to care about ARPA-E Open data? (Ask David)
  - No. There will be some TERRA REF data to be processed in the future though.
- [x] How much do we have to care about Clowder (if at all)? (Ask David)
  - To the extent that Clowder would be useful, but it should be 'at the edges', like other databases (including Bety)

## References

- <https://cctools.readthedocs.io/en/latest/makeflow/#garbage-collection>
- <https://sci-f.github.io/>
- Prior art/alternative to SCIF: <https://biocontainers.pro/#/multipackage>
- Tool metadata: <https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html>

**Comment by julianpistorius** [comment]
# SCIF notes & references

## Spec

https://github.com/sci-f/sci-f.github.io/blob/master/spec/spec-v1.md

## Goals

https://sci-f.github.io/goals#what-scif-is-not

> SCIF handles the step that comes before any kind of workflow manager is used, and is best used alongside some kind of container technology. It works with, and does not replace either of those things.

https://sci-f.github.io/goals#modules

>  - **Internal**: A second common scenario might be a single container that holds executables to perform different steps of a pipeline, perhaps so that the researcher can use the same container to run multiple steps, or perform any number of steps in parallel. This container would come with multiple *internal modules*, each performing a series of commands for one step in the pipeline (e.g., the step "mapping" uses internal commands from software `bwa` and `samtools`). The user doesn't need to know the specifics of the steps, but how to call them. We call this level "internal modules" because without any formal structure for the contents of containers, they are hidden, internal executables that must be found or described manually.
>  - **Development**: Containers can also serve modules that are represented at the ideal level for development. For this example, instead of providing the container as a node, or actions inside like "mapping", the smallest units of software are exposed, such as the executables `bwa` and `samtools`. It would be likely that a researcher developing a scientific pipeline would find this useful.

https://sci-f.github.io/goals#reproducible-practices

> # Reproducible Practices
> We have just discussed why internal modularity is important for container interaction, and this is also the case that it's useful for reproducibility. It is important to distinguish the entire container as a reproducible product, and different software modules inside of it that depend on being served through the container to ensure reproducibility. While the container itself is portable, and designed to contain all dependencies to support reproducibility, the SCi-F module in and of itself is not guaranteed to be. For example, a user might define a  module only with an `%apprun` section, implying that the folder only contains a runscript to execute. The user may have chosen to install dependencies for this script globally in the container, in the `%post` section, because perhaps they are shared across multiple modules. Under these conditions, if another user expected to add the module to a different build recipe, the dependencies from `%post` would be needed too. The host operating system also needs to be taken into consideration. A module with dependencies installed from the package manager "yum" would not move seamlessly into a debian base. However, appropriate checks and balances can be implemented into the process of moving applications:
> 
>  - For applications that must be portable outside of their initial container, users would be encouraged to include all dependency installs within the `%appinstall` section. If they were already installed during `%post`, the package would be found and skipped.
>  - Installing an application into a container would check for OS compatibility. This can be done automatically by storing information about the base OS with each application as a label. To encourage this practice, we have added a test and requirements of specifying one or more operating systems for any module contributed at <a href="https://sci-f.github.io/apps" target="_blank">https://sci-f.github.io/apps</a>. With these checks, we can have some confidence that the recipes for generating the apps are maximally portable.
> 
> Modular internal contents combined with reproducible portable environments via Singularity containers is a starting point for practicing good science.

https://sci-f.github.io/examples

> You might find SCIF useful if you:
> 
> ### A Research Scientist
> - want to package multiple environments or software modules to publish alongside a paper. A SCIF app might coincide with a particular step in your pipeline, and the entire analysis is run with a few calls to scientific filesystem entrypoints. You can share the scientific filesystem recipe for others to create, or (better) share a reproducible container built with it (that you used to conduct your analyses).

## Examples

Curated example applications: 

https://sci-f.github.io/apps/

Examples of installing a couple of relatively complicated workflow tools: 

- https://sci-f.github.io/apps/scif/workflow/workflow-cwltool/
- https://sci-f.github.io/apps/scif/workflow/workflow-toil/

Large number of bioinformatics recipes:

https://github.com/BU-ISCIII/scif_app_recipes

See an example of messing around with entry point - hopefully we don't have to do this (think it's necessary for native Docker integration in Nextflow), but just in case:

https://github.com/BU-ISCIII/scif_app_recipes/blob/master/Dockerfile_example

Their base image that speeds things up:

https://github.com/BU-ISCIII/CentOS7_base_image

## Detailed examples, including workflows

https://sci-f.github.io/apps/category/#Example

Snakemake and the Scientific Filesystem: 

https://sci-f.github.io/apps/examples/snakemake.scif

https://github.com/sci-f/snakemake.scif/blob/master/snakemake_tutorial.scif

(Above two links are useful for showing development vs production, and creating pipelines)

Singularity Scientific Workflow with SCI-F Example: 

https://sci-f.github.io/apps/examples/scientific

> SCI-F clearly has the advantage when it comes to a container creator embedding his or her work with implied metadata about software and container contents. SCI-F also makes it easier to package different run scripts with the container, and expose them easily to the user.

(Above post is really useful for evaluating capabilities of SCIF)

CarrierSeq: sequence analysis workflow with Singularity + SCIF: 

https://sci-f.github.io/apps/examples/carrierseq

https://sci-f.github.io/apps/examples/carrierseq.scif

(Above two links are useful for showing development vs production, and creating pipelines)

## App generator

https://sci-f.github.io/apps/generate/

## Cookiecutter template

https://github.com/vsoch/pipeline-template

## Each app can be in its own `.scif` file

Example lines from a Dockerfile:

```
RUN scif install /code/hello-world.scif
RUN scif install /code/bye-world.scif
```

## CWL and SCIF?

https://github.com/vsoch/scif/issues/49

## Nextflow and SCIF

https://github.com/vsoch/scif/issues/47

https://github.com/BU-ISCIII/nextflow-scif 

Be sure to read the notes section about how to layer the workflow engine & SCIF - it might be relevant if we want to use the native Docker integration with Makeflow: 

https://github.com/BU-ISCIII/nextflow-scif#notes

This might be a good way to package Makeflow and the transformers in the same container:

https://github.com/BU-ISCIII/nextflow-scif/blob/feature/nextflow-scif-alt/README-ALT.md

(I think we basically want a Makeflow version of this template above)

https://github.com/vsoch/rnatoy.scif

## First steps

- Follow these instructions: https://sci-f.github.io/builder/

**Comment by julianpistorius** [comment]
# Makeflow

## Install from Conda

https://cctools.readthedocs.io/en/latest/install/#install-from-conda

## Remote files

https://cctools.readthedocs.io/en/latest/makeflow/#mounting-remote-files

**Comment by julianpistorius** [comment]
# Plan (WIP)

Inspired by: https://github.com/BU-ISCIII/scif_app_recipes

## Transformers

Each transformer should live in its own Git repository.

A transformer repository should have:

- `.scif` recipe (including descriptive metadata and `%apptest` sections)
- Test data sufficient to test the correctness of the algorithm
  - For a small amount of data that rarely changes, include the data directly in the repository
  - For larger test data sets use something like Git Large File Storage, Git Annex, or Dat
  - See: https://drivendata.github.io/cookiecutter-data-science/#data-is-immutable
- Dockerfile
    - Based on a standard image
    - Maybe https://github.com/SCiO-systems/cgspatial-notebook
    - Or make our own: https://github.com/jupyter/cookiecutter-docker-stacks
- A Jupyter notebook with scaffolding code for development and testing of an algorithm
    - Ideally this notebook can be converted to a script and run as the canonical way of running the algorithm
- Some way to precisely specify library dependencies (and sub-dependencies) exhaustively and precisely (e.g. [Pipenv](https://pipenv.pypa.io/en/latest/) and [Conda environment.yaml](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html))
- A file that implements a way to perform [Continuous Analysis](https://greenelab.github.io/continuous_analysis/) on the algorithm (using TravisCI, CircleCI, Drone, etc.) with the bundled test data (Optional: As well as externally supplied data)
- Links in the README.md file to launch the Jupyter notebook in an interactive workbench environment like [Binder](https://mybinder.org/) and [CyVerse VICE](https://learning.cyverse.org/projects/vice/en/latest/getting_started/about.html)
- Instructions for how to use the Docker image as a reliable and reproducible development environment (especially when combined with PyCharm)

There should be a [cookiecutter](https://github.com/cookiecutter/cookiecutter) recipe to bootstrap a new transformer with all the requirements above.

## Workflows

Each workflow should live in its own Git repository.

A workflow repository should have:

- A 'scif_app_recipes' directory
- [Git submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) for each transformer/step in the workflow, in the 'scif_app_recipes' directory, checked out with a specific hash (or tag)
- A [`.jx` file](https://cctools.readthedocs.io/en/latest/jx/jx/) to describe the workflow
- Test data sufficient to test the correctness of the pipeline
  - For a small amount of data that rarely changes, include the data directly in the repository
  - For larger test data sets use something like Git Large File Storage, Git Annex, or Dat
  - See: https://drivendata.github.io/cookiecutter-data-science/#data-is-immutable
- A Dockerfile
    - Based on a standard image (see notes for Transformers above)
    - `RUN` line to install CCTools/Makeflow as a SCIF app
    - `RUN` lines to install transformers and any required utility apps as SCIF apps using the `.scif` recipes in the submodules. E.g.
    ```    
    COPY ./scif_app_recipes/*  /opt/
    RUN scif install /opt/makeflow-scif/makeflow.scif
    RUN scif install /opt/utility-app-01/utility-app-01.scif
    RUN scif install /opt/transformer-01/transformer-01.scif
    RUN scif install /opt/transformer-02/transformer-02.scif
    ```
    - Find inspiration from [this Dockerfile](https://github.com/BU-ISCIII/scif_app_recipes/blob/master/Dockerfile_example)
- A file that implements a way to perform [Continuous Analysis](https://greenelab.github.io/continuous_analysis/) on the workflow (using TravisCI, CircleCI, Drone, etc.) with the bundled test data (Optional: As well as externally supplied data)

There should be a [cookiecutter](https://github.com/cookiecutter/cookiecutter) recipe to bootstrap a new workflow with all the requirements above.

**Comment by Chris-Schnaufer** [comment]
I think this looks good. I have some comments and questions below.

Completely agree with using a cookie cutter approach to bootstrapping a new workflow!

Some more comments:
- I would prefer to stay away from conda since, in my experience, it is **really** heavyweight and therefore should be avoided. So far everything that I've ever needed to be done can be done with pip.
- Workflows should be able to be executed through an API - no front end aside from listening service
- IMO bundled test data should be available outside of a repo,. since not everyone will want it, updating test files should be easy outside of source control (use version numbers), increases download times of repos, test files should be accessible outside of source control, CI should be always against the latest dataset & code (as defined by SC (source control) push & merge)

Some questions:
1) If we provide a docker image with a development environment do we need cookiecutter? Something that can just be used without any steps other than launching would be ideal (for VICE, Binder, local to machine, etc).

Depending upon anything arising from my C's & Q's, how soon can you put together a demo using what we have for canopy cover workflow?

**Comment by julianpistorius** [comment]
Hi @Chris-Schnaufer! Thanks for the feedback.

> - I would prefer to stay away from conda since, in my experience, it is really heavyweight and therefore should be avoided. So far everything that I've ever needed to be done can be done with pip.

Personally I prefer pip as well, but Conda is very popular with researchers. It also allows installing non-Python libraries (R, etc.) for algorithm developers using languages other than Python. 

See here for a comparison: https://www.anaconda.com/understanding-conda-and-pip/

Idea: We could ask the developer to choose between Conda and Pipenv during the Cookiecutter process when creating a new transformer.

> - Workflows should be able to be executed through an API - no front end aside from listening service

I think the API service is outside the scope of this task, and a different concern. I see it as a convenience 'value add'. It should always be possible to execute a workflow manually, on the command line, without a backing service. 

Imagine someone wanting to run a workflow on their laptop, offline, next to the field where the drone is flying. Or a researcher in the future wanting to reproduce somebody's workflow that is archived as an executable artifact in a data repository.

A service like you describe can be built to consume GitHub repositories (and Docker images) which conform to the standard above.

Moreover what you're describing is essentially the Continuous Analysis as a Service product that CyVerse is building, and which we could leverage once basic workflows work.

> - IMO bundled test data should be available outside of a repo,. since not everyone will want it, updating test files should be easy outside of source control (use version numbers), increases download times of repos, test files should be accessible outside of source control, CI should be always against the latest dataset & code (as defined by SC (source control) push & merge)

I agree, to some extent. This is what I was thinking:

https://drivendata.github.io/cookiecutter-data-science/#data-is-immutable

_"If you have a small amount of data that rarely changes, you may want to include the data in the repository."_

For larger test data sets it would be preferable to use something like Git Large File Storage, Git Annex, or dat. I'll update the wording. 

> If we provide a docker image with a development environment do we need cookiecutter? Something that can just be used without any steps other than launching would be ideal (for VICE, Binder, local to machine, etc).

Yes, we need Cookiecutter. It is not an alternative to a development environment. Cookiecutter is a way to create a new repository/project. Think of it as a more powerful mechanism than the GitHub 'template repositories' we currently have for new transformers/algorithms. 

See here: https://github.com/cookiecutter/cookiecutter

_"A command-line utility that creates projects from cookiecutters (project templates), e.g. creating a Python package project from a Python package project template."_

The algorithm or pipeline developer would only run the `cookiecutter` command line tool once - when they create a new repository (transformer or workflow) for the first time. This new repository will include a Dockerfile (which can be used to build an image for local development) as well as a README.md file with a link to launch the repository in something like Binder or VICE.

Does that make sense?

> Depending upon anything arising from my C's & Q's, how soon can you put together a demo using what we have for canopy cover workflow?

I can have a demo by the end of next week.

Please let me know if anything is still unclear, or if you have suggestions for improvements.

**Comment by Chris-Schnaufer** [comment]
> Idea: We could ask the developer to choose between Conda and Pipenv during the Cookiecutter process when creating a new transformer.

Good idea. We could start with the one and add the other (conda) depending upon difficulty with creating them

> It should always be possible to execute a workflow manually,

Agreed: I just don't want to have only a Jupyter (or GUI) front end:

>a Dockerfile (which can be used to build an image for local development) as well as a README.md file with a link to launch the repository in something like Binder or VICE.
>
>Does that make sense?

Yes it makes sense.

I can't wait to see the demo

**Comment by julianpistorius** [comment]
## Work log

### 2020-04-13

- [x] Fork <https://github.com/Chris-Schnaufer/drone-makeflow> and clone locally
- [x] Find first transformer
    - See `sample.canopycover_workflow.yml`
    - ODM workflow apparently, then soil mask
- [x] Check out ODM workflow
    - [x] Q: What image does it use?
        - A: `agdrone/transformer-opendronemap:2.0`
        - See `sample.canopycover_workflow.yml` -> `workflow[0].docker_image`
        - <https://hub.docker.com/r/agdrone/transformer-opendronemap>
    - [x] Q: Where is the Git repo & Dockerfile for ODM image?
        - A: <https://github.com/AgPipeline/transformer-opendronemap>
    - [x] Q: What is the base Docker image for this image?
        - A: `FROM opendronemap/odm:0.9.1`
    - [x] Q: What is the `opendronemap/odm:0.9.1` image based on?
        - A: `FROM phusion/baseimage`
        - Note: Installs `gdal` here
        - [x] Q: Will it be compatible with other transformers?
            - A: Yes, I think so. See below at `phusion/baseimage`
        - DockerHub: <https://hub.docker.com/r/opendronemap/odm>
        - DockerHub Dockerfile: <https://hub.docker.com/r/opendronemap/odm/dockerfile>
        - GitHub repo: <https://github.com/OpenDroneMap/ODM>
    - [x] Q: What is the `phusion/baseimage` image based on
        - A: Ubuntu 18.04
        - `ARG BASE_IMAGE=ubuntu:18.04`
        - _"Ubuntu, plus modifications for Docker-friendliness, and solves the PID 1 zombie reaping problem"_
        - DockerHub: <https://hub.docker.com/r/phusion/baseimage>
        - GitHub repo: <https://github.com/phusion/baseimage-docker>
        - Dockerfile: <https://github.com/phusion/baseimage-docker/blob/master/image/Dockerfile>
    - [x] Q: Is `phusion/baseimage` compatible with other transformers?
        - A: Should be. It is based on Ubuntu 18.
- [x] Check out soil mask workflow
    - [x] Q: What image does it use?
        - A: `agdrone/transformer-soilmask:2.0`
        - See `sample.canopycover_workflow.yml` -> `workflow[1].docker_image`
        - <https://hub.docker.com/r/agdrone/transformer-soilmask>
    - [x] Q: Where is the Git repo & Dockerfile for ODM image?
        - A: <https://github.com/AgPipeline/drone-pipeline-environment>
        - Seems all the other transformers (besides ODM) live in the same repo
        - But the code for the transformers are submodules. For example the
        code for soilmask: <https://github.com/AgPipeline/transformer-soilmask>
    - [x] Q: What is the soilmask image based on?
        - A: `agdrone/drone-base-image:1.2`
        - Note: `gdal` and stuff installed as part of `drone-base-image`
    - [x] Q: What is the `agdrone/drone-base-image:1.2` image based on?
        - A: `agpipeline/base-image:1.3`
        - See <https://github.com/AgPipeline/drone-pipeline-environment/blob/master/base-transformer-class/Dockerfile>
        installed
    - [x] Q: What is the `agpipeline/base-image:1.3` image based on?
        - A: `FROM ubuntu:18.04`
        - See <https://github.com/AgPipeline/base-docker-support>
        - <https://github.com/AgPipeline/base-docker-support/blob/master/base-image/Dockerfile>
        - Note: `base-image` is where
            - user `extractor` created
            - Python 3 installed

#### Extra notes about ODM and `phusion/baseimage`

- ODM image is based on Ubuntu 18 via `phusion/baseimage`
    - Added benefit of allowing multiple users
    - See: <https://github.com/phusion/baseimage-docker#wait-i-thought-docker-is-about-running-a-single-process-in-a-container>
    - We should consider using this for transformer base image
    - See: <https://github.com/phusion/baseimage-docker#using-baseimage-docker-as-base-image>
    - Might mean a developer could run whole pipeline, with all workers, in single
    _CONTAINER_ (not single image and multiple containers)

#### Concerns & unknowns

- Not sure that ODM and other transformers will happily live in same image
    - ODM installs & uses Python 2, for example
    - [x] Test by creating a `.scif` file for ODM and soil mask
    - Alternatives if above does not work:
        - Base image for rest of transformers on ODM (not ideal)
        - Loosen single-image requirement
        - Multi-stage build that copies ODM stuff to right place in `/scif`
- Not sure `phusion/baseimage` would work with Singularity
    - [ ] Test `phusion/baseimage` in Singularity
    - [ ] Test `opendronemap/odm` in Singularity

**Comment by julianpistorius** [comment]
## Work Log

### 2020-04-14

#### Package ODM in SCIF

- [x] Fork & clone <https://github.com/AgPipeline/drone-pipeline-environment>
- [x] Fork & clone <https://github.com/AgPipeline/transformer-opendronemap>
- [x] Create scif branch
- [x] Create new Dockerfile based on `phusion/baseimage`
- [x] Build container `agdrone/transformer-opendronemap:3.0`
- [x] Add python, pip & `scif` to Dockerfile
    - Python 3.5.2 already installed
- [x] Install test SCIF recipe Dockerfile and test
- [x] Create SCIF recipe based on <https://hub.docker.com/r/opendronemap/odm/dockerfile>
- [x] Change to multi-stage Dockerfile
    - [x] Copy the `/code` directory from `opendronemap/odm:0.9.1` to run stage
        - `COPY --from=opendronemap/odm:0.9.1 /code /`
    - [x] `COPY --from=builder /usr/local/lib/python2.7/dist-packages /usr/local/lib/python2.7/dist-packages`
    - [x] Figure out which operating system packages actually need to be installed
        - Almost done.
- [x] Build two containers for ODM:
    - [x] Official
    - [x] From SCIF recipe
- [-] Compare OpenDroneMap functionality between two containers.
    - Test data:
        - <https://github.com/OpenDroneMap/ODM/blob/master/odm_docker_readme.txt>
        - <https://github.com/OpenDroneMap/ODMdata>
    - [x] Official
        - [x] Just run
    - [x] SCIF recipe
        - [x] Just run
        - [-] Failed first time. Try again. Might have had to do because I ran
        it at the same time as the official one.
        - Got it to work, but not sure it's giving the right output. Will leave
        for later
    - [ ] Compare console output of two runs
    - [ ] Compare output results of two runs (output/official vs output/scif)

Get the test data:

```bash
cd ~/docker_bind_root
git clone --depth 1 https://github.com/yjmenezes/OpenDroneMap.git
cd OpenDroneMap/tests/test_data
```

Run official container:

```bash
OUTPUT_DIR=$(pwd)/output/offical
mkdir -p "${OUTPUT_DIR}"
rm -r "${OUTPUT_DIR}/odm_*" "${OUTPUT_DIR}/opensfm"
docker run -it --rm \
    -v "$(pwd)/gcp_list.txt:/code/gcp_list.txt" \
    -v "$(pwd)/images:/code/images" \
    -v "${OUTPUT_DIR}/opensfm:/code/opensfm" \
    -v "${OUTPUT_DIR}/odm_meshing:/code/odm_meshing" \
    -v "${OUTPUT_DIR}/odm_georeferencing:/code/odm_georeferencing" \
    -v "${OUTPUT_DIR}/odm_orthophoto:/code/odm_orthophoto" \
    -v "${OUTPUT_DIR}/odm_texturing:/code/odm_texturing" \
    opendronemap/odm:0.9.1 --mesh-size 100000
```

Run SCIF container:

```bash
OUTPUT_DIR=$(pwd)/output/scif
mkdir -p "${OUTPUT_DIR}"
rm -r "${OUTPUT_DIR}/odm_*" "${OUTPUT_DIR}/opensfm"
docker run -it --rm \
    -v "$(pwd)/gcp_list.txt:/code/gcp_list.txt" \
    -v "$(pwd)/images:/code/images" \
    -v "${OUTPUT_DIR}/opensfm:/code/opensfm" \
    -v "${OUTPUT_DIR}/odm_meshing:/code/odm_meshing" \
    -v "${OUTPUT_DIR}/odm_georeferencing:/code/odm_georeferencing" \
    -v "${OUTPUT_DIR}/odm_orthophoto:/code/odm_orthophoto" \
    -v "${OUTPUT_DIR}/odm_texturing:/code/odm_texturing" \
    agdrone/transformer-opendronemap:3.0 run odm --mesh-size 100000
```

Later:

- [ ] Copy `/code` to `/scif/apps/odm`? Or leave it where it is and use
    entry point fiddling?
- [ ] Idea: Copy `/usr/lib/x86_64-linux-gnu/` from official image to SCIF one
    - Put it in the `/scif/` directory for odm ap
    - Set `LD_LIBRARY_PATH` appropriately
        - <https://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html>
        - <https://stackoverflow.com/questions/13428910/how-to-set-the-environmental-variable-ld-library-path-in-linux>
        - <https://stackoverflow.com/questions/47682750/why-changing-ld-library-path-has-no-effect-in-ubuntu?noredirect=1&lq=1>
        - `export LD_DEBUG=all`
    - Alternatively look at which dynamic libraries the odm binaries link to and install those packages
    - Or just give up and install all the stupid `-dev` packages from the original image, they're going to go to
      multi-stage image builds at some point anyway
- [ ] Use a separate Python environment for ODM (and each app)
    - Test if that currently happens
- [ ] Put Python dependencies for ODM in a Pipfile and install that way (by copying from requirements.txt)
- [ ] Read these to see if possible to grab layers from official ODM image
    - <https://dille.name/blog/2018/08/19/how-to-reduce-the-build-time-of-a-monolithic-docker-image/>
    - <https://dille.name/blog/2018/09/07/how-to-automate-the-merging-of-layers-from-docker-images-in-powershell/>
    - Figure out myself: "How to Fix Package Manager Databases when Merging Layers from Docker Images (coming soon)"
- [ ] What happens when you COPY from two build images into a single location?
    - [ ] When the set of files don't overlap?
    - [ ] When the set of files overlap?
    - [ ] When the source & destination directories have the same name? Do they
    get clobbered?

**Comment by julianpistorius** [comment]
@Chris-Schnaufer which branch of AgPipeline/transformer-soilmask should I be basing my work on?

https://github.com/AgPipeline/transformer-soilmask/branches

**Comment by julianpistorius** [comment]
### 2020-04-15

#### Package Soil Mask in SCIF (1)

- [x] Fork repo, clone & create new branch
- [x] Create new Dockerfile based on `phusion/baseimage`
    - See ODM one I made
- [x] Build Docker image `agdrone/transformer-soilmask:3.0`
    - `docker build -t agdrone/transformer-soilmask:3.0 .`
- [x] Add pip & `scif` to Dockerfile
- [x] Run Docker container
    - `docker run --rm agdrone/transformer-soilmask:3.0 `
- [-] Add transformer boilerplate code to repo
    - [x] Use custom build outputs and a new Dockerfile based on
    `agdrone/transformer-soilmask:2.0` to get boilerplate code locally
      ```bash
      export DOCKER_BUILDKIT=1
      docker build --output type=local,dest=test -f boilerplate.Dockerfile .
      ```
    - Later: Submodule? Or Python package? Or just generated by Cookiecutter?
    - Or can it all just be handled by recipe.scif `%runapp` entrypoint?

Notes:

- Docker [custom build outputs](https://github.com/docker/cli/blob/master/docs/reference/commandline/build.md#custom-build-outputs)
    could be useful to see what will be in an image (or diff between builds)
- Build with URL, repo, or tar context could be also be useful
    - <https://github.com/docker/cli/blob/master/docs/reference/commandline/build.md#build-with-url>
- Have a look at buildx: Docker CLI plugin for extended build capabilities with BuildKit
    - <https://github.com/docker/buildx>
    - Might it be useful when combined with SCIF?
- Idea: Use `chroot` inside Docker container, combined with SCIF?
    - So should be able to install different versions of the same operating
    system package?
    - And might also solve Python library issue without virtual environments.

**Comment by julianpistorius** [comment]
### 2020-04-20

#### Package Soil Mask in SCIF (2)

- [x] Add transformer boilerplate code to repo
    - `transformer.py` is different between master and the Docker image. Try different branch.
    - Pinged Chris: <https://github.com/AgPipeline/issues-and-projects/issues/132#issuecomment-614288037>
    - Trying `develop` branch
    - Still different. Looks like the Docker image has not been rebuilt with
    latest code in develop (at least 2.0)
    - [x] Try with `agdrone/transformer-soilmask:latest`
        ```bash
        mkdir -p test/2.0
        mv test/* test/2.0/
        mkdir -p test/latest
        mv boilerplate.Dockerfile boilerplate-2.0.Dockerfile
        cat boilerplate-2.0.Dockerfile | sed -E 's/2\.0/latest/' > boilerplate-latest.Dockerfile
        export DOCKER_BUILDKIT=1
        docker build --output type=local,dest=test/latest -f boilerplate-latest.Dockerfile .
        ```
      - No. Still the same. And same digest as 2.0, so I should have known.
    - [x] Try using Chris's fork
        - Nope. Doesn't look like there is a fork by him
    - [x] Check out `drone-pipeline-environment/common_transformers` branch
        - Ahah! Yes. Barring some linting differences these are the same:
            - `transformer-soilmask/test/2.0/home/extractor/transformer_class.py`
            - `drone-pipeline-environment/base-transformer-class/transformer_class.py`
        - Also:
            - `transformer-soilmask/test/2.0/home/extractor/transformer.py`
            - `drone-pipeline-environment/transformer-soilmask/soilmask-code/transformer`
        - The hash for the `drone-pipeline-environment/transformer-soilmask/soilmask-code`
        submodule is `3f5bba5`
        - The `3f5bba5` corresponds to the tip of the `agpipeline/typo_fixes` branch
        - [x] Therefore rebase my `scif` branch over the `typo_fixes` branch
    - [x] Copy `drone-pipeline-environment/base-transformer-class/transformer_class.py`
    to `transformer-soilmask/transformer_class.py` project
        ```bash
        cp ../drone-pipeline-environment/base-transformer-class/transformer_class.py transformer_class.py
        ```
    - [x] Get `entrypoint.py`
        - Check <https://github.com/AgPipeline/base-docker-support>
        - `master` branch, `32bdf46c5b9af446fa35138da0d4daccba7af6a7` hash
        - Following pretty much the same except for a comment in the former:
            - `base-docker-support/base-image/entrypoint.py`
            - `transformer-soilmask/test/2.0/home/extractor/entrypoint.py`
        ```bash
        cp ../base-docker-support/base-image/entrypoint.py entrypoint.py
        ```
- [x] Create app section in recipe.scif for soilmask
    - [x] In Dockerfile copy files over to `/scif/apps/soilmask/src`
    - [x] Use `agdrone/transformer-opendronemap:3.0` as base for now
        - My connection is too slow to download things
- [-] Build image
    ```bash
    docker build -t agdrone/transformer-soilmask:3.0 .
    ```
    - [ ] Install requirements
        - Use `%appenv` to specify a Python virtual environment
    - [ ] Test, fix, iterate till `entrypoint.py` works

Later:

- [ ] Automate `transformer_class.py` & `entrypoint.py` copy with `curl`? To keep it updated in the future?
    - Or submodule? Or Python package? Or just generated by Cookiecutter?
    - What are they doing?
    - `transformer_class.py`: importing `terrautils` and using `lemnatec`
    sensor metadata
    - This seems odd for a drone pipeline
- [ ] For development workflow and containers:
    - <https://sci-f.github.io/apps/examples/carrierseq.scif>
    - <https://github.com/amojarro/carrierseq>

**Comment by julianpistorius** [comment]
### 2020-04-21

#### Package Soil Mask in SCIF (3)

- [-] Build image
    ```bash
    docker build -t agdrone/transformer-soilmask:3.0 .
    ```
    - [-] Install requirements
    - [-] Test, fix, iterate till `entrypoint.py` works
        - [x] scikit-image
        - [x] opencv
        - [x] PyYAML
        - [ ] terrautils
- Conda:
    ```bash
    conda install --prefix "${PWD}" scikit-image opencv PyYAML
    conda list --export --prefix "${PWD}" > src/package-list.txt
    ```

**Comment by Chris-Schnaufer** [comment]
Perhaps the following might help understand how transformers are built:
1) entrypoint.py is built into a base docker image
2) transfomer_class.py is next built into a new image
3) transformer.py is the built into newest image
The plan is to put entrypoint.py into a library or BUILD Dockerfile command and do the same for the tranformer_class.py file

**Comment by julianpistorius** [comment]
Thanks for the clarification @Chris-Schnaufer. 🙇‍♂️

**Comment by julianpistorius** [comment]
### 2020-04-23

#### Package Soil Mask in SCIF (4)

- [x] Build image
    ```bash
    docker build -t agdrone/transformer-soilmask:3.0 .
    ```
    - [x] Install requirements
    - [x] Test, fix, iterate till `entrypoint.py` works
        - [x] terrautils
- [x] Check terrautils dependencies, install via Conda

**Comment by julianpistorius** [comment]
### 2020-04-24

#### Combine ODM & Soil Mask in one recipe/container

- [x] Base new branch (scif) on the [docker_support](https://github.com/AgPipeline/drone-makeflow/tree/docker_support) branch of AgPipeline/drone-makeflow repo
    - See <https://github.com/julianpistorius/drone-makeflow/tree/scif>
- [x] Submodules
    - [x] ODM: julianp/scif
    - [x] soilmask : julianp/scif
- [x] Dockerfile
    - [x] ODM recipe
    - [x] soilmask recipe
- [x] Build combined image
```bash
docker build -f docker/Dockerfile -t drone/makeflow:3.0 .
```
- [x] Run scif and apps
    - [x] odm
    - [x] soil mask
```
$ docker run -it --rm drone/makeflow:3.0 apps
       odm
  soilmask
$ docker run -it --rm drone/makeflow:3.0 help soilmask
    This app provides an entrypoint to the soilmask tool
    docker run drone/makeflow:3.0 run soilmask
$ docker run -it --rm drone/makeflow:3.0 run soilmask
[soilmask] executing /bin/bash /scif/apps/soilmask/scif/runscript
ERROR:root:No metadata paths were specified.
ERROR:root:Stopping processing
WARNING:root:Writing result to a file was requested but a file path wasn't provided.
WARNING:root:    Skipping writing to a file.

{
  "error": "No metadata paths were specified.",
  "code": -1
}
$ docker run -it --rm drone/makeflow:3.0 shell soilmask
[soilmask] executing /bin/bash
root@08b5cb2978b7:/scif/apps/soilmask# pwd
/scif/apps/soilmask
root@08b5cb2978b7:/scif/apps/soilmask# exit
exit
```

**Comment by julianpistorius** [comment]
### 2020-04-24

#### Add CCTools back to drone/makeflow container

- [x] `scif/makeflow/recipe.scif`

```
$ docker run -it --rm drone/makeflow:3.1 apps
  soilmask
  makeflow
       odm
$ docker run -it --rm drone/makeflow:3.1 run makeflow
[makeflow] executing /bin/bash /scif/apps/makeflow/scif/runscript
ERROR conda.cli.main_run:execute(32): Subprocess for 'conda run ['makeflow']' command failed.  (See above for error)
makeflow: No makeflow specified and file "./Makeflow" could not be found.
makeflow: Run "makeflow -h" for help with options.
```

**Comment by julianpistorius** [comment]
### 2020-05-01

#### Add transformer boilerplate code to ODM SCIF recipe #139

<https://github.com/AgPipeline/issues-and-projects/issues/139>

- [x] Dockerfile in transformer-opendronemap repo:
    - [x] Copy all boilerplate source over
    - [x] Change entrypoint in scif recipe to be `entrypoint.py
        - Also added parameters: `exec python src/entrypoint.py "${@}"`
    - [x] Build
    - [x] Test & repeat :arrow_heading_up:
        - Fails due to python2
        - [x] Install miniconda
        - [x] Install requirements using conda
        - [x] Export environment.yml
- [x] Dockerfile in drone-makeflow repo:
    - [x] Copy all the source over
    - [x] Build
    - [x] Test & repeat :arrow_heading_up:
- [x] Modify soil-mask recipe to have parameters: `"${@}"`

- [x] Run `apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*` in recipes & Dockerfiles

- [ ] Check sizes of image layers
    - Combine the conda environments?
    - Or strip transformers down to bare 'scientific packages' (check Jacob's progress) and jettisoning the legacy stuff (`pyclowder`, etc.)

**Comment by julianpistorius** [comment]
### 2020-05-04

#### Call two SCIF transformers using Makeflow

_Note: Didn't make much progress. Tried to reverse-engineer ODM transformer
without much luck._

- [ ] Makeflow .jx file to call ODM
    - [ ] Step back and try to build Chris's container directly
        - Then run it the way the instructions say
        - Reverse engineer from there
        - Then...

    - [ ] Convert `canopycover_workflow.yaml` to `workflow_configuration.json`
        - The JSON file is the configuration file for Makeflow
        - Actually, there's almost nothing there.
        - See `sample.canopycover_workflow.yml`
        - There might be some dynamic additions in `run_workflow.sh`

- [ ] Makeflow .jx file to call soil mask
- [ ] Check output

**Comment by julianpistorius** [comment]
### 2020-05-05

#### Back to basics

Make my own MVP of transformers (ODM, soil mask, canopy cover), without current
boilerplate. Just has to illustrate my plan (pseudo code), not be complete,
correct, etc.

#### ODM

From David:

- Just generate an orthomosaic
- Skip gcp_list if not provided
- Also skip opensfm, odm_meshing

- [x] Run official container on Chris's test images:

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow/test
OUTPUT_DIR=$(pwd)/output/offical
mkdir -p "${OUTPUT_DIR}"
rm -r "${OUTPUT_DIR}/odm_*" "${OUTPUT_DIR}/opensfm"
docker run --rm \
    -v "$(pwd)/images:/code/images" \
    -v "${OUTPUT_DIR}/opensfm:/code/opensfm" \
    -v "${OUTPUT_DIR}/odm_meshing:/code/odm_meshing" \
    -v "${OUTPUT_DIR}/odm_georeferencing:/code/odm_georeferencing" \
    -v "${OUTPUT_DIR}/odm_orthophoto:/code/odm_orthophoto" \
    -v "${OUTPUT_DIR}/odm_texturing:/code/odm_texturing" \
    opendronemap/odm:0.9.1
```

- [x] Make new container with new odm-plain recipe
    - [x] Modify SCIF of ODM to not use boilerplate
    - [x] Add as separate recipe
    - [x] New temp-odm-plain.Dockerfile
    - [x] Build container

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow
docker build -f docker/temp-odm-plain.Dockerfile -t drone/makeflow:3.4 .
docker run --rm drone/makeflow:3.4 apps
```

Should show:

```
       odm
  makeflow
 odm-plain
  soilmask
```

- [x] Run `odm-plain` app on Chris's test images:

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow/test
OUTPUT_DIR=$(pwd)/output/scif
mkdir -p "${OUTPUT_DIR}"
rm -r "${OUTPUT_DIR}/odm_*" "${OUTPUT_DIR}/opensfm"
docker run --rm \
    -v "$(pwd)/images:/code/images" \
    -v "${OUTPUT_DIR}/opensfm:/code/opensfm" \
    -v "${OUTPUT_DIR}/odm_meshing:/code/odm_meshing" \
    -v "${OUTPUT_DIR}/odm_georeferencing:/code/odm_georeferencing" \
    -v "${OUTPUT_DIR}/odm_orthophoto:/code/odm_orthophoto" \
    -v "${OUTPUT_DIR}/odm_texturing:/code/odm_texturing" \
    drone/makeflow:3.4 run odm-plain
```

- [x] Remove transformer-opendronemap submodule
- [x] Modify `drone-makeflow/docker/Dockerfile` to use new odm recipe
for now
- [x] Build `drone/makeflow:3.5` with new odm recipe

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow
docker build -f docker/Dockerfile -t drone/makeflow:3.5 .
docker run --rm drone/makeflow:3.5 apps
```

- [x] Run `odm` app on Chris's test images:

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow/test
OUTPUT_DIR=$(pwd)/output/scif-02
mkdir -p "${OUTPUT_DIR}"
rm -r "${OUTPUT_DIR}/odm_*" "${OUTPUT_DIR}/opensfm"
docker run --rm \
    -v "$(pwd)/images:/code/images" \
    -v "${OUTPUT_DIR}/opensfm:/code/opensfm" \
    -v "${OUTPUT_DIR}/odm_meshing:/code/odm_meshing" \
    -v "${OUTPUT_DIR}/odm_georeferencing:/code/odm_georeferencing" \
    -v "${OUTPUT_DIR}/odm_orthophoto:/code/odm_orthophoto" \
    -v "${OUTPUT_DIR}/odm_texturing:/code/odm_texturing" \
    drone/makeflow:3.5 run odm
```


#### Soil Mask

- [x] Modify `transformer-soilmask/recipe` to only install `numpy` and `opencv`
- [x] Grab new `/scif/apps/soilmask/src/environment.yml`

Next: 

- [ ] Make the minimum required Python boilerplate around the `gen_plant_mask` method in soil mask transformer
- [ ] Basic tests for soil mask
- [ ] Write basic .jx to chain odm and soil mask

#### Notes

- These look interesting:
    - <https://github.com/RicardoRodriguezArg/crop_row_detection>
    - <https://github.com/its4land>
        - <https://github.com/its4land/docker-fly-and-create>
- Also Tyson's Dockerfile for VICE:
    - <https://github.com/tyson-swetnam/vice-node-odm>

**Comment by julianpistorius** [comment]
### 2020-05-06

#### Soil Mask

- [x] Make the minimum required Python boilerplate around the `gen_plant_mask` method in soil mask transformer
    - See [HackMD: Mimimal Canopy Cover Pipeline](https://hackmd.io/mSYV8WPsQ-GSsg84rCfxmA)
    - Code: <https://github.com/julianpistorius/transformer-soilmask/tree/07038dc88efe5b098500f31d6a5b16172620b38a>
- [ ] Basic tests for soil mask
- [ ] Write basic .jx to chain odm and soil mask

**Comment by julianpistorius** [comment]
### 2020-05-07

Code: https://github.com/julianpistorius/drone-makeflow/tree/bc06d106bd9a302065a785c27fcaf901d4cfa03f

#### Makeflow to chain ODM & soil mask

- [x] Fix soil mask Dockerfile FROM
- [x] Update soil mask submodule in drone-makeflow repo

```bash
git submodule update --remote -f --init --recursive
git add transformer-soilmask
git commit -m "Updated transformer-soilmask submodule"
git push
```

- [x] Build new drone-makeflow container

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow
export DOCKER_BUILDKIT=1
docker build -f docker/Dockerfile -t drone/makeflow:3.6 .
docker run --rm drone/makeflow:3.6 apps
```

- [x] Run apps to sanity check

```bash
docker run --rm drone/makeflow:3.6 run soilmask
docker run --rm drone/makeflow:3.6 run odm
```

- [x] Rebuild image till works

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow
export DOCKER_BUILDKIT=1
docker build -f docker/test.Dockerfile -t drone/makeflow:3.7 .

docker run --rm drone/makeflow:3.7 run soilmask
docker run --rm drone/makeflow:3.7 run odm
```

- [x] Run `soilmask --help` and add that to the `%apphelp` section
- [x] Build new makeflow image

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow
export DOCKER_BUILDKIT=1
docker build -f docker/test.Dockerfile -t agpipeline/scif-drone-pipeline:1.0 .
```

- [x] Run odm transformer, in container, with `scif`, with test data
    - [x] Use `--project-path` and `project name` with `SCIF_APPDATA`

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow
DATA_DIR=$(pwd)/test/data
\rm -rf $DATA_DIR
# Make the data directories for each app
docker run --rm agpipeline/scif-drone-pipeline:1.0 apps | xargs -L 1 -I % mkdir -p "$DATA_DIR/%"
mkdir -p "$DATA_DIR/odm/images"
cp $(pwd)/test/images/*.JPG "$DATA_DIR/odm/images/"

docker run -it --rm \
    -v "${DATA_DIR}:/scif/data" \
    agpipeline/scif-drone-pipeline:1.0 shell odm

# Then in Docker shell:
# Cheating... a bit, to make odm work with SCIF conventions
ODM_PROJECT_NAME=$SCIF_APPNAME
ODM_PROJECT_PATH=$SCIF_DATA
scif run odm $ODM_PROJECT_NAME --project-path $ODM_PROJECT_PATH
```

- [x] Run soilmask transformer with test data
    - [x] Use:
        - `SCIF_APPDATA=/scif/data/soilmask`
        - `SCIF_APPDATA_odm=/scif/data/odm`

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow
DATA_DIR=$(pwd)/test/data

docker run -it --rm \
    -v "${DATA_DIR}:/scif/data" \
    agpipeline/scif-drone-pipeline:1.0 shell soilmask

# Then in Docker shell:
scif run soilmask $SCIF_APPDATA_odm/odm_orthophoto/odm_orthophoto.tif --output-file $SCIF_APPDATA/soilmask-output.npz
```

- [x] Write basic .jx to chain odm and soil mask

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow
DATA_DIR=$(pwd)/test/data
\rm -rf $DATA_DIR
# Make the data directories for each app
docker run --rm agpipeline/scif-drone-pipeline:1.0 apps | xargs -L 1 -I % mkdir -p "$DATA_DIR/%"
mkdir -p "$DATA_DIR/odm/images"
cp $(pwd)/test/images/*.JPG "$DATA_DIR/odm/images/"

docker run -it --rm \
    -v "${DATA_DIR}:/scif/data" \
    -v "$(pwd)/workflow.jx:/scif/apps/makeflow/src/workflow.jx" \
    -v "$(pwd)/jx-args.json:/scif/apps/makeflow/src/jx-args.json" \
    agpipeline/scif-drone-pipeline:1.0 shell makeflow

# Then in Docker shell:
conda run --prefix "${SCIF_APPROOT}/conda" makeflow --help

conda run --prefix "${SCIF_APPROOT}/conda" jx2json \
    --args "${SCIF_APPROOT}/src/jx-args.json" \
    --pretty \
    "${SCIF_APPROOT}/src/workflow.jx" \
    | tee "${SCIF_APPDATA}/workflow.json"

conda run --prefix "${SCIF_APPROOT}/conda" makeflow \
    --jx \
    --jx-args="${SCIF_APPROOT}/src/jx-args.json" \
    --log-verbose \
    --retry-count=1 \
    --change-directory="${SCIF_APPDATA}" \
    --makeflow-log="${SCIF_APPDATA}/workflow.jx.makeflowlog" \
    --batch-log="${SCIF_APPDATA}/workflow.jx.batchlog" \
    "${SCIF_APPROOT}/src/workflow.jx"
```

- [x] Modify makeflow recipe to do all the stuff above
- [x] Move `ODM_PROJECT_NAME` and `ODM_PROJECT_PATH` to `%apprun odm` section
    - That way there is no problem with shell escaping issues
    - Simpler as well
- [x] Delete superfluous files from drone-makeflow repo

- [x] Rebuild image with latest recipes

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow
export DOCKER_BUILDKIT=1
docker build -f docker/Dockerfile -t agpipeline/scif-drone-pipeline:1.1 .
```

- [x] Run new image

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow
DATA_DIR=$(pwd)/test/data
\rm -rf $DATA_DIR
# Make the data directories for each app
docker run --rm agpipeline/scif-drone-pipeline:1.1 apps | xargs -L 1 -I % mkdir -p "$DATA_DIR/%"
mkdir -p "$DATA_DIR/odm/images"
cp $(pwd)/test/images/*.JPG "$DATA_DIR/odm/images/"

docker run --rm \
    -v "${DATA_DIR}:/scif/data" \
    agpipeline/scif-drone-pipeline:1.1 run makeflow

```

Didn't quite work.

- [x] Try again:
    - Apparently multiple execs don't work, so added a jx2json entrypoint

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow
export DOCKER_BUILDKIT=1
docker build -f docker/test.Dockerfile -t agpipeline/scif-drone-pipeline:1.2 .
```

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow
DATA_DIR=$(pwd)/test/data
\rm -rf $DATA_DIR
# Make the data directories for each app
docker run --rm agpipeline/scif-drone-pipeline:1.2 apps | xargs -L 1 -I % mkdir -p "$DATA_DIR/%"
mkdir -p "$DATA_DIR/odm/images"
cp $(pwd)/test/images/*.JPG "$DATA_DIR/odm/images/"

docker run --rm \
    -v "${DATA_DIR}:/scif/data" \
    agpipeline/scif-drone-pipeline:1.2 run makeflow

```

- [x] From scratch, once more, with feeling:

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow
export DOCKER_BUILDKIT=1
docker build -f docker/Dockerfile -t agpipeline/scif-drone-pipeline:1.1 .
```

```bash
# Make the data directories for each app
docker run --rm agpipeline/scif-drone-pipeline:1.1 apps | xargs -L 1 -I % mkdir -p "$DATA_DIR/%"
mkdir -p "$DATA_DIR/odm/images"
cp $(pwd)/test/images/*.JPG "$DATA_DIR/odm/images/"

docker run --rm \
    -v "${DATA_DIR}:/scif/data" \
    agpipeline/scif-drone-pipeline:1.1 run makeflow
```

Later:

- [ ] Readme & instructions (How simple can you make the first run?)
    - [ ] `jx-args.json`
    - [ ] Docker build
    - [ ] Getting test images (Bake test images into image? Only 68MB.)
    - [ ] Run container with bind mounts
        - Separate bind mounts for each app?
    - [ ] Makeflow clean: Intermediate as well as all artifacts
    - [ ] Development flow (e.g. PyCharm + Docker)
- [ ] Basic tests for soil mask
- [ ] Generate jx-arg.json from scif environment variables?
    - Jinja templating?
    - Hard-code certain things which won't change into workflow.jx?
- [ ] Maybe use `makeflow --argv=`
- [ ] Install 'soil mask' Python package from GitHub
- [ ] Use `--wrapper=` for makeflow recipe (maybe `scif shell`?)

**Comment by julianpistorius** [comment]
### 2020-05-08

#### SCIF + Makeflow demo prep

Goal: Should only need three commands to build and run pipeline from scratch

- Git clone
- Docker build
- Docker run

- [x] Do apt stuff in Dockerfile, not odm recipe
    - Because not HPC friendly
- [x] Add conda to the path as part of Dockerfile
- [x] Flatten recipes folder? Like `scif_app_recipes`?
- [x] Don't use Git submodule for soil mask
    - Clone repo as part of recipe (or download .tar.gz for now)
    - Future: conda/pip install
- [x] Move Dockerfile to root of repo, and remove `docker` directory
- [x] Install miniconda in /usr/local or /opt
    - <https://conda.io/projects/conda/en/latest/user-guide/install/macos.html#install-macos-silent>
- [x] Separate build stages for each app recipe to speed dev builds
    - Kinda almost got there. Will get back to this.

```bash
cd ~/dev/diag/AgPipeline/drone-makeflow
export DOCKER_BUILDKIT=1

docker build --progress=plain --target install_miniconda -t agpipeline/scif-drone-pipeline:install_miniconda .
docker build --progress=plain --target base_scif -t agpipeline/scif-drone-pipeline:base_scif .
docker build --progress=plain --target makeflow_scif -t agpipeline/scif-drone-pipeline:makeflow_scif .
docker build --progress=plain --target odm_base -t agpipeline/scif-drone-pipeline:odm_base .
docker build --progress=plain --target soilmask_scif -t agpipeline/scif-drone-pipeline:soilmask_scif .
docker build --progress=plain --target odm_scif -t agpipeline/scif-drone-pipeline:odm_scif .
docker build --progress=plain --target combined_scif -t agpipeline/scif-drone-pipeline:combined_scif .
docker build --progress=plain -t agpipeline/scif-drone-pipeline:1.3 .
```

```bash
mkdir -p test
wget -O test/testimages.tar https://de.cyverse.org/dl/d/84A57A62-B6EB-4826-ADC4-337D4A0ABBEA/images.tar
tar xf images.tar -C test/
INPUT_DATA_DIR=$(pwd)/test/images
OUTPUT_DATA_DIR=$(pwd)/test/output
mkdir -p "${OUTPUT_DATA_DIR}"
docker run --rm \
    -v "${INPUT_DATA_DIR}:/scif/data/odm/images" \
    -v "${OUTPUT_DATA_DIR}:/scif/data/soilmask" \
    agpipeline/scif-drone-pipeline:1.3 run makeflow
```

- [x] Make app data directories as first step in workflow
- [x] Hard-code jx-arg.json values in workflow.jx
    - [x] Test

Later:

- [ ] Download test data as second step in workflow
    - Hash/checksum
- [ ] What happens when you try to install a recipe twice?
    - Answer: Problems. But might be the key to multi-stage build per recipe.
    - Keep pushing.
- [ ] Rename transformer-soilmask repo to soilmask
- [ ] Basic tests for soil mask
- [ ] Export/expose volumes in Dockerfile for app data directories


#### Demo with David: Notes

What's missing?

- [ ] Expose `/scif/data/<app>` for each app as a volume in the Dockerfile
    - <https://docs.docker.com/engine/reference/builder/#volume>
    - [ ] Modify Docker run instructions to mount those locations. Otherwise intermediate artifacts of the workflow steps will be lost after the workflow completes.
- [ ] Tests
    - [ ] Integration tests, end to end
    - [ ] Tests for each algorithm
    - [ ] Use `%apptest`
    - [ ] Sanity tests/quality control gates
- [ ] Install soil mask as a Python/Conda package
    - Instead of downloading the code from my repo
    - Can `pip install` from GitHub as first approach
- [ ] Readme for soil mask repo
- [ ] Metadata in SCIF recipes
- [ ] Figure out how to use a different Docker build stage for each SCIF app
    - When the build stages are combined you should still end up with a working SCIF system with all the apps listed
- [ ] Make ODM deterministic (getting different results each time)
- [ ] Documentation
    - Diagrams
    - Screencast
- [ ] Document development process (PyCharm, Docker, Conda, etc.) for:
    - Algorithm developer
    - Pipeline developer
- [ ] Cookiecutter recipes
    - Algorithm/scientific package
    - Pipeline

**Comment by julianpistorius** [comment]
Relevant pull requests:

- <https://github.com/AgPipeline/drone-makeflow/pull/1>
- <https://github.com/AgPipeline/transformer-soilmask/pull/12>

**Comment by dlebauer** [comment]
what more needs to be done here? Can some of the discussion be captured in documentation?

**Comment by dlebauer** [comment]
Todo: see comment above

**Comment by julianpistorius** [comment]
I'll have a look.

**Comment by dlebauer** [comment]
closing since this seems complete, over 2.5 years old
